File: backend/core/week_calendar.py
# -*- coding: utf-8 -*-
"""Единая логика недельного календаря (настраиваемый первый день недели)."""
from __future__ import annotations

from datetime import date, timedelta

# Python weekday: понедельник=0 … воскресенье=6
WEEKDAY_MON = 0
WEEKDAY_TUE = 1
WEEKDAY_WED = 2
WEEKDAY_THU = 3
WEEKDAY_FRI = 4
WEEKDAY_SAT = 5
WEEKDAY_SUN = 6

DEFAULT_WEEK_START_DAY = WEEKDAY_SAT

WEEKDAY_LABELS_RU: dict[int, str] = {
    WEEKDAY_MON: "Понедельник",
    WEEKDAY_TUE: "Вторник",
    WEEKDAY_WED: "Среда",
    WEEKDAY_THU: "Четверг",
    WEEKDAY_FRI: "Пятница",
    WEEKDAY_SAT: "Суббота",
    WEEKDAY_SUN: "Воскресенье",
}


def normalize_week_start_day(value: int | None) -> int:
    if value is None:
        return DEFAULT_WEEK_START_DAY
    v = int(value)
    if v < 0 or v > 6:
        return DEFAULT_WEEK_START_DAY
    return v


def week_start_for_date(d: date, start_day: int = DEFAULT_WEEK_START_DAY) -> date:
    """Начало недели, содержащей дату d."""
    sd = normalize_week_start_day(start_day)
    delta = (d.weekday() - sd + 7) % 7
    return d - timedelta(days=delta)


def week_start_iso(iso_date: str, start_day: int = DEFAULT_WEEK_START_DAY) -> str:
    d = date.fromisoformat(str(iso_date)[:10])
    return week_start_for_date(d, start_day).isoformat()


def week_dates_from_anchor(anchor_date: str, start_day: int = DEFAULT_WEEK_START_DAY) -> list[str]:
    start = week_start_for_date(
        date.fromisoformat(str(anchor_date)[:10]),
        start_day,
    )
    return [(start + timedelta(days=i)).isoformat() for i in range(7)]


def is_weekday_in_week(day_iso: str, weekday: int, start_day: int = DEFAULT_WEEK_START_DAY) -> bool:
    """True если day_iso — заданный день недели в неделе с start_day."""
    d = date.fromisoformat(str(day_iso)[:10])
    ws = week_start_for_date(d, start_day)
    return (d - ws).days == weekday


def is_sunday_in_sat_week(day_iso: str) -> bool:
    """Воскресенье в неделе с субботы (legacy food rules)."""
    return is_weekday_in_week(day_iso, WEEKDAY_SUN, WEEKDAY_SAT)


def week_number(iso_date: str, start_day: int = DEFAULT_WEEK_START_DAY) -> int:
    d = date.fromisoformat(str(iso_date)[:10])
    return int(d.isocalendar()[1])


def previous_week_range(
    on_date: date | None = None,
    start_day: int = DEFAULT_WEEK_START_DAY,
) -> tuple[date, date]:
    """Календарная неделя, непосредственно предшествующая неделе с on_date."""
    d = on_date or date.today()
    cur_start = week_start_for_date(d, start_day)
    prev_end = cur_start - timedelta(days=1)
    prev_start = prev_end - timedelta(days=6)
    return prev_start, prev_end

File: frontend/src/shared/utils/weekCalendar.ts
/** Единая логика недельного календаря (Python weekday: пн=0 … вс=6). */

export const WEEKDAY_MON = 0;
export const WEEKDAY_TUE = 1;
export const WEEKDAY_WED = 2;
export const WEEKDAY_THU = 3;
export const WEEKDAY_FRI = 4;
export const WEEKDAY_SAT = 5;
export const WEEKDAY_SUN = 6;

export const DEFAULT_WEEK_START_DAY = WEEKDAY_SAT;

export const WEEKDAY_OPTIONS: { value: number; label: string }[] = [
  { value: WEEKDAY_MON, label: "Понедельник" },
  { value: WEEKDAY_TUE, label: "Вторник" },
  { value: WEEKDAY_WED, label: "Среда" },
  { value: WEEKDAY_THU, label: "Четверг" },
  { value: WEEKDAY_FRI, label: "Пятница" },
  { value: WEEKDAY_SAT, label: "Суббота" },
  { value: WEEKDAY_SUN, label: "Воскресенье" },
];

function pythonWeekday(d: Date): number {
  const js = d.getDay();
  return js === 0 ? 6 : js - 1;
}

export const WEEKDAY_SHORT_RU = ["Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Вс"] as const;

export function weekdayLabelsFromStart(startDay = DEFAULT_WEEK_START_DAY): string[] {
  const sd = normalizeWeekStartDay(startDay);
  return Array.from({ length: 7 }, (_, i) => WEEKDAY_SHORT_RU[(sd + i) % 7]);
}

/** Число пустых ячеек перед 1-м числом месяца (Python weekday: пн=0 … вс=6). */
export function monthGridLeadingPadding(
  year: number,
  month: number,
  startDay = DEFAULT_WEEK_START_DAY,
): number {
  const sd = normalizeWeekStartDay(startDay);
  const first = new Date(year, month, 1);
  return (pythonWeekday(first) - sd + 7) % 7;
}

export function normalizeWeekStartDay(value: number | null | undefined): number {
  if (value == null || value < 0 || value > 6) return DEFAULT_WEEK_START_DAY;
  return value;
}

export function weekStartForDate(isoDate: string, startDay = DEFAULT_WEEK_START_DAY): string {
  const sd = normalizeWeekStartDay(startDay);
  const d = new Date(`${isoDate.slice(0, 10)}T12:00:00`);
  const delta = (pythonWeekday(d) - sd + 7) % 7;
  d.setDate(d.getDate() - delta);
  return d.toISOString().slice(0, 10);
}

export function weekDatesFromAnchor(isoDate: string, startDay = DEFAULT_WEEK_START_DAY): string[] {
  const start = new Date(`${weekStartForDate(isoDate, startDay)}T12:00:00`);
  return Array.from({ length: 7 }, (_, i) => {
    const x = new Date(start);
    x.setDate(start.getDate() + i);
    return x.toISOString().slice(0, 10);
  });
}

export function shiftWeekStart(weekStart: string, weeks: number): string {
  const d = new Date(`${weekStart.slice(0, 10)}T12:00:00`);
  d.setDate(d.getDate() + weeks * 7);
  return d.toISOString().slice(0, 10);
}

export function formatWeekLabel(weekStartIso: string): string {
  const ws = new Date(weekStartIso.slice(0, 10) + "T12:00:00");
  const we = new Date(ws);
  we.setDate(we.getDate() + 6);
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${pad(ws.getDate())}.${pad(ws.getMonth() + 1)} – ${pad(we.getDate())}.${pad(we.getMonth() + 1)}.${we.getFullYear()}`;
}
